### ISSUE TITLE
Next.js 13 Amplify - CustomerError: The size of the build output exceeds the max allowed size of 230686720 bytes

### ISSUE THREAD

**Issue by jitendra-koodo** [aws-amplify/amplify-hosting#3863]
### Before opening, please confirm:

- [X] I have checked to see if my question is addressed in the [FAQ](https://github.com/aws-amplify/amplify-hosting/blob/master/FAQ.md).
- [X] I have [searched for duplicate or closed issues](https://github.com/aws-amplify/amplify-hosting/issues?q=is%3Aissue+).
- [X] I have read the guide for [submitting bug reports](https://github.com/aws-amplify/amplify-hosting/blob/master/CONTRIBUTING.md).
- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.
- [X] I have removed any sensitive information from my code snippets and submission.

### App Id

new_app

### AWS Region

us-east-1

### Amplify Hosting feature

Frontend builds

### Frontend framework

Next.js

### Next.js version

13

### Next.js router

App Router

### Describe the bug

https://github.com/aws-amplify/amplify-hosting/issues/3647

Issue above has been abruptly closed without providing any details .. Creating this new ticket 

200MB limit is still there : why can it not be removed?

### Expected behavior

Build should not fail

### Reproduction steps

Build
Front end build steps fails ;:

### Log output

<details>

```
# Put your logs below this line
2023-12-19T07:14:02.836Z [INFO]: # Caching completed
2023-12-19T07:14:19.932Z [ERROR]: !!! CustomerError: The size of the build output (236302771) exceeds the max allowed size of230686720 bytes. Please reduce the size ofyour build output and try again.

```

</details>

**Comment by jitendra-koodo** [comment]
2023-12-19T07:14:02.836Z [INFO]: # Caching completed
2023-12-19T07:14:19.932Z [ERROR]: !!! CustomerError: The size of the build output (236302771) exceeds the max allowed size of230686720 bytes. Please reduce the size ofyour build output and try again.

**Comment by jitendra-koodo** [comment]
https://github.com/aws-amplify/amplify-hosting/issues/3647

Issue above has been abruptly closed without providing any details .. Creating this new ticket

200MB limit is still there : why can it not be removed?

**Comment by jitendra-koodo** [comment]
Can anyone suggest how we can move out of this Amplify to some other service which does not have this limit? Any recommendation? Our production builds are failing and we need to find alternatives very quickly. Business loss. Thanks.

**Comment by mauerbac** [comment]
I'm sorry to hear you are experiencing this issue. We made changes last week that removed non-critical known to be large packages to help stay below the 200MB limit. 

Can you look at your output and breakdown the different packages and let me know? Running `du -csh .next/*`should tell us that.

**Comment by victorccccc** [comment]
Hi 👋 , sorry for the inconvenience. We added logic to remove unneeded dependency files that are typically large (e.g some unused heavy binaries under `@esbuild` and `@swc` packages).

To review your build output folder, you can:

1. Run following command
```
aws amplify get-job --app-id ${APP_ID} --branch-name ${BRANCH_NAME} --job-id ${BUILD_ID}
```
2.Find artifactsUrl (presigned S3 url) under job -> steps -> stepName: "BUILD"
3. Download the file which is your build output.

The bundle size issue could happen if your app has heavy dependencies. You can check size of the dependency under the `node_modules` folder in the build output. You could try to use a lighter weight dependency or remove some of the files you wouldn't need.

**Comment by jitendra-koodo** [comment]
Thanks @victorccccc @mauerbac for quick revert..

App ID: dkc2twse2pm4r
Job url#/dkc2twse2pm4r/staging-in/1

==
2023-12-19T06:41:23.553Z [INFO]: # Executing command: du -sh *
2023-12-19T06:41:23.556Z [INFO]: 4.0K	amplify.sh
2023-12-19T06:41:23.556Z [INFO]: 4.0K	amplify.yml
                                 28K	apidata.json
                                 48K	blog-related-data.json
                                 0	__cache.tar
2023-12-19T06:41:23.556Z [INFO]: 548K	components
2023-12-19T06:41:23.556Z [INFO]: 4.0K	context
                                 4.0K	data.json
                                 0	eslintrc.js
                                 88K	fonts
                                 44K	helpers
                                 4.0K	jsconfig.json
                                 8.0K	lib
                                 8.0K	middleware.js
                                 36K	next.config.js
2023-12-19T06:41:23.653Z [INFO]: 424M	node_modules
                                 4.0K	package.json
2023-12-19T06:41:23.653Z [INFO]: 280K	pages
                                 24K	public
                                 4.0K	README.md
                                 12K	scripts
                                 4.0K	seo-hreflangs.json
                                 72K	services
                                 4.0K	sort.json
2023-12-19T06:41:23.654Z [INFO]: 456K	styles
                                 200K	yarn.lock
2023-12-19T06:41:23.654Z [INFO]: # Executing command: du -h -d1 public
2023-12-19T06:41:23.656Z [INFO]: 4.0K	public/.well-known
                                 24K	public
2023-12-19T06:41:23.656Z [INFO]: # Executing command: du -h -d1 .next/static
2023-12-19T06:41:23.657Z [INFO]: 2.0M	.next/static/chunks
2023-12-19T06:41:23.657Z [INFO]: 8.0K	.next/static/fFlvAR4iijcKnCOhx6OaR
                                 344K	.next/static/css
                                 2.3M	.next/static
2023-12-19T06:41:23.658Z [INFO]: # Executing command: du -h -d1 .next/server
2023-12-19T06:41:23.659Z [INFO]: 28M	.next/server/chunks
2023-12-19T06:41:23.662Z [INFO]: 166M	.next/server/pages
2023-12-19T06:41:23.662Z [INFO]: 194M	.next/server
2023-12-19T06:41:23.662Z [INFO]: # Executing command: du -h -d1 .next/cache
2023-12-19T06:41:23.663Z [INFO]: 72K	.next/cache/eslint
2023-12-19T06:41:23.664Z [INFO]: 0	.next/cache/swc
                                 280M	.next/cache/webpack
                                 280M	.next/cache
2023-12-19T06:41:23.664Z [INFO]: # Completed phase: build
2023-12-19T06:41:23.669Z [INFO]: ## Build completed successfully
2023-12-19T06:41:23.670Z [INFO]: # Starting caching...
2023-12-19T06:41:23.674Z [INFO]: # Creating cache artifact...
2023-12-19T06:41:39.686Z [INFO]: # Created cache artifact
2023-12-19T06:41:39.852Z [INFO]: # Uploading cache artifact...
2023-12-19T06:41:42.793Z [INFO]: # Uploaded cache artifact
2023-12-19T06:41:42.793Z [INFO]: # Caching completed
2023-12-19T06:41:59.624Z [ERROR]: !!! CustomerError: The size of the build output (236302652) exceeds the max allowed size of230686720 bytes. Please reduce the size ofyour build output and try again.
2023-12-19T06:41:59.731Z [INFO]: # Starting environment caching...

**Comment by jitendra-koodo** [comment]
I tried to do some basic maths to find out upper limit on number of pages which could be released on my website with this 200MB limit.

I roughly measured storage when I had 70 pages (case 1) and 700 pages (case2) (each page similar size in KB) . In case 1, total size was ~95MB. In case 2, total size was ~175MB. So if there is basic overhead of size A and per page need size B, then 

A + 70B = 95
A + 700B = 175

B = ~120KB (it matches with the size visible on Chrome)

This means, no matter what optimization we do, at some point when page count go up, it'll break 200MB barrier even when all other things are optimized.

 Our page count will double in next one week, What's the solution here?

**Comment by irvan-photon** [comment]
It's also happen in Next.js 14

**Comment by gzManuel** [comment]
There is another duplicated [issue](https://github.com/aws-amplify/amplify-hosting/issues/3841), @jitendra-koodo  I found two solutions

1. Upgrade to **Nextjs 14** and update the build image of amplify to **Amazon Linux:2023**
2. Delete the deployed branch of amplify, remove package-lock.json and generate it again, update the build image of amplify to **Amazon Linux:2023** and deploy the branch again.

**Comment by ggorge-etiqa** [comment]
Another little update is that `rm -rf node_modules/next/next-swc-fallback` is only really needed.

**Comment by jitendra-koodo** [comment]
Point is that even with all these small optimizations, at some point you are going to hit the 200MB limit and at that point you have to move out of Amplify

**Comment by ggorge-etiqa** [comment]
@jitendra-koodo I know this is frustrating, but every solution has it's own limits. 
Docs report this build output limit here: https://docs.aws.amazon.com/amplify/latest/userguide/troubleshooting-ssr-deployment.html#build-output-too-large but I think it should be also mentioned there: https://docs.aws.amazon.com/amplify/latest/userguide/quotas-chapter.html

**Comment by jitendra-koodo** [comment]
@ggorge-etiqa Not sure what you mean.. which other solution has 200MB limit? moreover this limit is not shown anywhere on console at any point of time. People go to that doc only after hitting this limit.

**Comment by dannypaz** [comment]
Rebump. I ran into this bug with Nextjs 14 and a very basic starter app. After using https://github.com/aws-amplify/amplify-hosting/issues/3863#issuecomment-1893933629 I was able to resolve

**Comment by andre-mr** [comment]
> Another little update is that `rm -rf node_modules/next/next-swc-fallback` is only really needed.
> 
> I'm wondering why amplify is bothering on `node_modules` directory size when the `artifacts` folder is set to `.next`:
> 
> ```
>   artifacts:
>     baseDirectory: .next
>     files:
>       - "**/*"
> ```

where did you added those lines? at the end of build phase? i tried only this line and got an error related to swc at the end of "cloning git" step.

don't know if it was a coincidence and a temp error, but in a desperate way i changed my build file to this:

version: 1
frontend:
  phases:
    preBuild:
      commands:
        - rm -rf .next
        - rm -rf node_modules
        - npm i
    build:
      commands:
        - npm run build
        - rm -rf .next/standalone/node_modules/next/next-swc-fallback
        - rm -rf node_modules/next/next-swc-fallback
  artifacts:
    baseDirectory: .next
    files:
      - '**/*'
  cache:
    paths:
      - node_modules/**/*

and it worked, lol. in the next update i'll try to change this, today we are very hurry here.

by the way: my local .next folder, after cleaned up and made from next build is 199mb size. if bundle limit is 200 it's at the limit anyway, but i couldn't find anything unused to remove from dependencies or imports.
AND i don't know if amplify is counting node_modules (another huge folder) in this 200mb size limit (if it's counting, it would never pass because i would have almost 400mb size .next+node_modules).

**Comment by alexHewittProcter** [comment]
I'm also experiencing this issue even after upgrading to Next 14! 
I've tried several different attempts including changing the base image I use, removing specific dependencies and attempting to remove the cache that exists between builds

Is there any update from AWS on this? I've wasted several hours trying to fix this problem and I have to question why we're being limited to 200mb for our artefacts when the service quotas allow for 5GB?

**Comment by nicolasleguizamon** [comment]
Hello, 
We have same issue! 
Next: 13.2.4

**Comment by ekpono** [comment]
I am having the same issue. This is annoying

Next: 14.0.4
```
[ERROR]: !!! CustomerError: The size of the build output (306216844) exceeds the max allowed size of230686720 bytes. Please reduce the size ofyour build output and try again.
```

**Comment by guy-a** [comment]
Adding `rm -rf node_modules/canvas` to the build (amplify.yml) worked for me.

i.e.
```yaml
version: 1
frontend:
  phases:
    preBuild:
      commands:
        - npm ci
    build:
      commands:
        - npm run build
        - rm -rf node_modules/canvas
```

**Comment by ekpono** [comment]
> Adding `rm -rf node_modules/canvas` to the build (amplify.yml) worked for me.
> 
> i.e.
> 
> ```yaml
> version: 1
> frontend:
>   phases:
>     preBuild:
>       commands:
>         - npm ci
>     build:
>       commands:
>         - npm run build
>         - rm -rf node_modules/canvas
> ```

I did the same think, I'm guessing having jest pulls in canvas which is about 100MB+

**Comment by pachoclo** [comment]
Running into this issue for a second time. Can we please get a clear reason why the 200mb limit is still being enforced? We really don't want to have to switch to a different hosting service...

**Comment by OussamaFadlaoui** [comment]
Bump

**Comment by OussamaFadlaoui** [comment]
Weird details on fixing this issue for me:
 * Whilst trying to debug the problem, I found out that switching to NPM made the build go through successfuly;
 * Eager to continue using PNPM, I experimented with PNPM's `node-linker` [config](https://pnpm.io/npmrc#node-linker) and set it to `node-linker=hoisted`. This means no symlinks are created in `node_modules`, and resulted in a successful build using PNPM.

What is weird, is that whilst I was debugging the failed builds, I would download the artefact files and they would amount to around ~130 MB. I just downloaded a ZIP-file of the artefacts on a successful build using PNPM, and the total size amounted to 138M. This is really vague, and I hope this extra info sheds light or helps someone with the same issue.

I do not feel comfortable continuing using Amplify, but I love the idea of it. I just wish it wouldn't be such a rocky start.

**Comment by GeuntaBuwono** [comment]
the solution should be use prod mode on install this one for example using yarn as installer 
```
    preBuild:
      commands:
        - yarn install --prod --pure-lockfile
```

and on package.json if you're using canvas as dependency you should add this 
```
	"resolutions": {
		"canvas": "link:./node_modules/.cache/null"
	},
```

**Comment by kgkg** [comment]
When using pnpm calculated build size is clearly wrong. 

Command `du -sh build` reports size of 151 MB, while amplify gives error message: `!!! CustomerError: The size of the build output (401160409) exceeds the max allowed size of 230686720 bytes.`.

It seems that Amplify is trying to calculate size by dereferencing symlinks (or hardlinks) that pnpm uses.

I was able to workaround this by using `cp -rL` and removing `node_modules/.pnpm` dir after that.

Here is log output of deploying with this workaround - notice the size of `flat_node_modules` directory before after removing `.pnpm` dir.
```
2024-10-18T07:32:25.967Z [INFO]: # Executing command: cp -rL standalone/node_modules flat_node_modules
2024-10-18T07:32:27.465Z [INFO]: # Executing command: du -sh flat_node_modules
2024-10-18T07:32:27.605Z [INFO]: 525M flat_node_modules
2024-10-18T07:32:27.605Z [INFO]: # Executing command: rm -fr flat_node_modules/.pnpm
2024-10-18T07:32:28.162Z [INFO]: # Executing command: du -sh flat_node_modules
2024-10-18T07:32:28.204Z [INFO]: 153M flat_node_modules
2024-10-18T07:32:28.205Z [INFO]: # Executing command: mv flat_node_modules my-app/build/compute/default/node_modules
2024-10-18T07:32:28.206Z [INFO]: # Executing command: rm -fr standalone
```

*Still 2 points:*
1) Amplify is incorrectly calculating build size when there are symlinks/hardlinks (pnpm states to use hardlinks). If this is not a bug then it's probably a feature...
2) Limit of 230 MB build size in current times, when only `node_modules` for a very simple app can weight 150M is plain ridiculous. Right now we're in situation in which we can't be sure if expanding our site will result with inability to even deploy it after some time. From business point of view this is unacceptable.

**Comment by kgkg** [comment]
> What is weird, is that whilst I was debugging the failed builds, I would download the artefact files and they would amount to around ~130 MB. I just downloaded a ZIP-file of the artefacts on a successful build using PNPM, and the total size amounted to 138M. This is really vague, and I hope this extra info sheds light or helps someone with the same issue.
> 
> I do not feel comfortable continuing using Amplify, but I love the idea of it. I just wish it wouldn't be such a rocky start.

That's very funny. I was thinking that failing to properly calculate size of build output when it includes symlinks/hardlinks is because they zip or copy the build content with derefencing links (similar to what `cp -L` does) but it turns out it's a plain bug 😄 

We're not comfortable with using Amplify either. Starter Remix app build takes ~150MB. How we can expand our site when there is risk we'll not be able to deploy it after some time? From business point of view this is unacceptable.

**Comment by kryota-dev** [comment]
I am creating a corporate blog site with Next.js v14 App Router.
The number of pages is quite huge (119 at the moment) and the response time from the headless CMS is not very fast, so I wanted to Static Generate at build time and defined a page component that would be a 'Dynamic Route' with `generateStaticParams` in the page component that would be the 'Dynamic Route' and deployed the build, I encountered the same problem.

```sh
!!! CustomerError: The size of the build output (279729806) exceeds the max allowed size of230686720 bytes. Please reduce the size ofyour build output and try again.
```

As we expect to have more articles in the future, we are finding it difficult to continue using Amplify 😢

**Comment by taraspos** [comment]
Hey @calavera, @Jay2113 

I have Next.js v14 application and facing the same problem with the build output at the moment.
If found previous replies mentioning that Next.js application shouldn't depend on `standalone` build output anymore which should resolve issues with hitting the limit:
- https://github.com/aws-amplify/amplify-hosting/issues/3601#issuecomment-1654668783
- https://github.com/aws-amplify/amplify-hosting/issues/3628#issuecomment-1665970652

However, inspecting build-time environment variables I can see that `NEXT_PRIVATE_STANDALONE=true`. Am I correct in assuming that currently standalone directory is used only to produce `node_modules`?

While it's not really documented how exactly Amplify puts together final deployment package, I was able to more or less reproduce the build locally following steps similar to the ones mentioned in FAQ[^1].

Bash script I came up with:
```bash
#!/bin/sh
export NEXT_PRIVATE_STANDALONE=true
yarn build

output_location=dist/apps/my_app/ # for non-monotrepo projects this will be just project root

# create directory to store our Amplify-like artifact
amplf_output_dir=$output_location/amplify_artifact
rm -rf $amplf_output_dir
amplf_compute_dir=$amplf_output_dir/compute/default
amplf_static_dir=$amplf_output_dir/static
mkdir -p $amplf_compute_dir
mkdir -p $amplf_static_dir

# copy node modules in resulting folder
cp -r $output_location/.next/standalone/node_modules $amplf_compute_dir/

# remove @swc and @esbuild from node modules
rm -rf "$amplf_compute_dir/node_modules/@swc"
rm -rf "$amplf_compute_dir/node_modules/@esbuild"

# copy .next dir
cp -r $output_location/.next $amplf_compute_dir/

# remove cache
rm -rf $amplf_compute_dir/.next/cache
# remove standalone dir, it was needed for node_modules only
rm -rf $amplf_compute_dir/.next/standalone

# move static files into separate diretory
mv $amplf_compute_dir/.next/static/* $amplf_static_dir

du -h -d1 $amplf_output_dir
```

[^1]: https://github.com/aws-amplify/amplify-hosting/blob/main/FAQ.md#measure-compute-apps-initializationstart-up-time-locally

**Comment by tylercod3r** [comment]
* same issue here w/ NextJS app v14. 
* tried amending the Amplify yaml build file, but no luck.
* have similar project using same import packages, but NO issues.  (i.e., it is also a Nextjs v14 app).
* for one built attempt it worked, and then after changing a couple integer values in code, this led to failed build process again (?).
* as temp work-around, seems that initial build works (e.g. I create a new Amplify app by connecting a Github repo.).
* so now, anytime I'd like to make repo updates, I have to create a fresh new Amplify app.   (= sad pandas.)
* hope this helps...

**Comment by aicam** [comment]
> Adding `rm -rf node_modules/canvas` to the build (amplify.yml) worked for me.
> 
> i.e.
> 
> ```yaml
> version: 1
> frontend:
>   phases:
>     preBuild:
>       commands:
>         - npm ci
>     build:
>       commands:
>         - npm run build
>         - rm -rf node_modules/canvas
> ```

I don't know how we can move up this answer but it worked perfectly for me and I tried many solutions here before finding this.

**Comment by chrissy-dev** [comment]
Just to chime in here, we're testing an Astro SSR app on Amplify at the moment and it's proving to be cumbersome due to this 200mb limit. However, I'm having some success by moving some modules to `devDependencies` and using `npm prune --production` on the build command. However, I echo the concerns of the many people above that the hard limit makes me question whether Amplify is the right choice moving forward due to no clear runway when the app grows.

This is a simple MVP test, and a minimal application - so I'm hesitant to dive any further into it.

Here's my `amplify.yml` for reference.  

```yml
version: 1
frontend:
    phases:
        preBuild:
            commands:
                - 'npm ci --cache .npm --prefer-offline'
        build:
            commands:
                - 'npm run build'
                - 'npm prune --production'
                - 'mv node_modules ./.amplify-hosting/compute/default'
    artifacts:
        baseDirectory: .amplify-hosting
        files:
            - '**/*'
    cache:
        paths:
            - '.npm/**/*'       
```

**Comment by iamrishan** [comment]
> Adding `rm -rf node_modules/canvas` to the build (amplify.yml) worked for me.
> 
> i.e.
> 
> version: 1
> frontend:
>   phases:
>     preBuild:
>       commands:
>         - npm ci
>     build:
>       commands:
>         - npm run build
>         - rm -rf node_modules/canvas

This works. if you are not using canvas just remove it by adding extra -rm command, canvas takes up half of the build size

**Comment by Shazil-55** [comment]
version: 1
frontend:
  phases:
    preBuild:
      commands:
        - npm ci
    build:
      commands:
        - npm run build
        - rm -rf node_modules/canvas


so removing the canvas I think it works for most of us 

I found out my last build which gets failed  and it was because I used canvas in that commit

**Comment by robinnepomukmai** [comment]
I had the same problem and for me building it with webpack instead of turbopack did the job

**Comment by socketbox** [comment]
_Not sure if this helps, but here is what I received from AWS support. The heavily indented text is from the application team engineers (or perhaps Tier 2/3 support):_


[A]. What directories are used to calculate the final build artifact size?

>> The overall idea is to follow the Amplify Deploymnet Specification: "https://docs.aws.amazon.com/amplify/latest/userguide/ssr-deployment-specification.html 
". There are open source adapters for this (e.g. Astro https://github.com/alexnguyennz/astro-aws-amplify 
), but for Next.js the adapter is internal to Amplify.

For SSR/compute apps (e.g., Next.js with WEB_COMPUTE platform), Amplify calculates the artifact size based on the compute bundle — a directory that Amplify's bundler assembles after your build completes. This is not your source directory or your raw build output; it's a separate staging directory that Amplify creates internally.

The compute bundle contains:

- Your Next.js build output (the .next/ directory, excluding the static/ and cache/ folders)
- Runtime dependencies (node_modules/ files required at runtime — more on this in [F])
- Your public/ folder contents
- .env and .env.production files
- A server runtime file and startup script
- Your package.json

The 220MB (230,686,720 bytes) limit is checked against the uncompressed total size of all files in this compute bundle directory. It is not the size of the zip file that gets uploaded.

Amplify's bundler independently determines which files to include in the compute bundle using Next.js trace files.


[B]. Reproducible and deterministic process to calculate artifact size locally:

Unfortunately, there is no straightforward way to exactly reproduce Amplify's artifact size calculation locally, because the compute bundle is assembled by Amplify's internal bundler. During the deploy phase. However, you can approximate it:

>> Run next build locally (with output: 'standalone' in your next.config.js if that's your configuration)

>> Examine the .nft.json trace files generated inside .next/server/ — these list every dependency file that will be included at runtime

>> The compute bundle is roughly the sum of:
  
  - All files referenced in the .nft.json trace files
  - The .next/ directory (minus static/ and cache/)
  - Your public/ folder
  - .env / .env.production
  - package.json
  - A small server runtime and startup script added by Amplify

>> Why the behavior is non-deterministic across branches: The trace files generated by next build depend on which pages/routes exist and what they import. Different branches may have:

  - Different dependencies or lock file resolutions
  - Different pages that import different modules
  - Different transitive dependency trees due to package manager hoisting

This means the set of traced node_modules files can vary between branches, causing some to land just above the 220MB limit and others just below.


[C]. Knowing the size of a successfully built artifact:

Currently, the artifact size is only displayed in the error message when the build fails.

On successful builds, the uncompressed size is logged 'INTERNALLY', but it is not surfaced to customers.


[D]. What exactly Amplify includes when calculating artifact size:

For a Next.js SSR app, the compute bundle that is measured against the 220MB limit includes:

>>  Traced runtime dependencies — not your entire node_modules/, but only the specific files identified as needed at runtime through Next.js file tracing (.nft.json files)

>> Next.js build output — the .next/ directory contents, excluding the static/ assets folder and the cache/ folder (static assets are bundled separately and do not count toward this limit
)

>> Public folder — files from your public/ directory

>> Environment files — .env and .env.production from your app root

>> Server runtime — a server entry point and startup script added by Amplify

>> package.json — from your project root

What is excluded:

>> .next/static/ (bundled separately as static assets)

>> .next/cache/

>> Build-time-only binaries like @esbuild platform binaries,@swc native binaries, and Next.js SWC fallback binaries

>> Dependencies traced to paths outside your project directory


[E]. How Amplify determines which node_modules/ files to include:

Amplify does not include your entire node_modules/ directory. It uses Next.js Node File Tracing (NFT) to determine the minimal set of files needed at runtime:

>> During next build, Next.js generates .nft.json trace files alongside each compiled page and route (e.g., .next/server/pages/index.js.nft.json). Each trace file lists the exact
dependency files that page needs at runtime.

>> Amplify's bundler reads all these trace files from .next/server/ and collects the union of all referenced dependency paths.

>> Additionally, Amplify traces the dependencies of its own server runtime file using @vercel/nft (the same tracing library Next.js uses internally).

>> The union of both sets (your app's traced dependencies + the server runtime's traced dependencies) is what gets copied into the compute bundle's node_modules/.

>> Certain patterns are excluded even if traced — specifically, platform-specific build-time binaries from @esbuild, @swc, and Next.js SWC fallback packages, which are incorrectly traced
by Next.js but not needed at runtime.
